SocketModeReceiver app process exits when any of its event listeners throws an exception

### Description

`SocketModeReceiver` does not handle the exceptions thrown in any of event listeners in a bolt-js app. This behavior is confusing to developers, plus it is inconsistent with other built-in receivers.

### What type of issue is this? (place an `x` in one of the `[ ]`)
- [x] bug
- [ ] enhancement (feature request)
- [ ] question
- [ ] documentation related
- [ ] example code related
- [ ] testing related
- [ ] discussion

### Requirements (place an `x` in each of the `[ ]`)
* [x] I've read and understood the [Contributing guidelines](https://github.com/slackapi/bolt/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

---

### Bug Report

Filling out the following details about bugs will help us solve your issue sooner.

#### Reproducible in:

package version: v3.0 and later
node version: any
OS version(s): any

#### Steps to reproduce:

1. Run the following Socket Mode app
2. Invite the app's bot user to a channel and mention the bot user
3. The bolt-js app process terminates

```javascript
const { App } = require("@slack/bolt");
const app = new App({
  socketMode: true,
  appToken: process.env.SLACK_APP_TOKEN,
  token: process.env.SLACK_BOT_TOKEN,
});

app.event("app_mention", async ({}) => {
  throw new Error("Something is wrong!");
});

(async () => {
  await app.start();
  console.log("⚡️ Bolt app is running!");
})();
```

#### Expected result:

The default error handler should print error logs and prevents the app process from terminating. Also, the error handler should be customizable like we do for other HTTP module based receivers.

#### Actual result:

The whole app process immediately terminates.

```
$ npx node socket-mode.js
⚡️ Bolt app is running!
[ERROR]  bolt-app UnknownError: Something is wrong!
    at asCodedError (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/errors.js:35:12)
    at App.handleError (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/App.js:588:57)
    at App.processEvent (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/App.js:573:25)
    at async SocketModeClient.<anonymous> (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/receivers/SocketModeReceiver.js:146:13) {
  code: 'slack_bolt_unknown_error',
  original: Error: Something is wrong!
      at /Users/kazuhiro.sera/github/bolt-starter/socket-mode.js:9:9
      at /Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/App.js:550:33
      at invokeMiddleware (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/middleware/process.js:20:16)
      at next (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/middleware/process.js:13:29)
      at Array.<anonymous> (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/middleware/builtin.js:233:15)
      at invokeMiddleware (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/middleware/process.js:12:53)
      at next (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/middleware/process.js:13:29)
      at Array.onlyEvents (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/middleware/builtin.js:63:11)
      at invokeMiddleware (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/middleware/process.js:12:53)
      at processMiddleware (/Users/kazuhiro.sera/github/bolt-starter/node_modules/@slack/bolt/dist/middleware/process.js:22:12)
}
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

$
```


#### Attachments:

Logs, screenshots, screencast, sample project, funny gif, etc.

--- a/examples/socket-mode/app.js
+++ b/examples/socket-mode/app.js
@@ -24,11 +24,12 @@ const clientOptions = {
 
 const app = new App({
   // receiver: socketModeReceiver,
-  token: process.env.BOT_TOKEN, //disable this if enabling OAuth in socketModeReceiver
+  token: process.env.SLACK_BOT_TOKEN, //disable this if enabling OAuth in socketModeReceiver
   // logLevel: LogLevel.DEBUG,
   clientOptions,
-  appToken: process.env.APP_TOKEN,
+  appToken: process.env.SLACK_APP_TOKEN,
   socketMode: true,
+  logLevel: LogLevel.DEBUG,
 });
 
 (async () => {
@@ -40,9 +41,9 @@ const app = new App({
 app.event('app_home_opened', async ({ event, client }) => {
   await client.views.publish({
     user_id: event.user,
-    view: { 
-      "type":"home",
-      "blocks":[
+    view: {
+      "type": "home",
+      "blocks": [
         {
           "type": "section",
           "block_id": "section678",
@@ -103,8 +104,7 @@ app.shortcut('launch_shortcut', async ({ shortcut, body, ack, context, client })
         ]
       }
     });
-  }
-  catch (error) { 
+  } catch (error) {
     console.error(error);
   }
 });
@@ -114,12 +114,44 @@ app.shortcut('launch_shortcut', async ({ shortcut, body, ack, context, client })
 // need app_mentions:read and chat:write scopes
 app.event('app_mention', async ({ event, context, client, say }) => {
   try {
-    await say({"blocks": [
+    await say({
+      "blocks": [
+        {
+          "type": "section",
+          "text": {
+            "type": "mrkdwn",
+            "text": `Thanks for the mention <@${event.user}>! Click my fancy button`
+          },
+          "accessory": {
+            "type": "button",
+            "text": {
+              "type": "plain_text",
+              "text": "Button",
+              "emoji": true
+            },
+            "value": "click_me_123",
+            "action_id": "first_button"
+          }
+        }
+      ]
+    });
+  } catch (error) {
+    console.error(error);
+  }
+});
+
+// subscribe to `message.channels` event in your App Config
+// need channels:read scope
+app.message('hello', async ({ message, say }) => {
+  // say() sends a message to the channel where the event was triggered
+  // no need to directly use 'chat.postMessage', no need to include token
+  await say({
+    "blocks": [
       {
         "type": "section",
         "text": {
           "type": "mrkdwn",
-          "text": `Thanks for the mention <@${event.user}>! Click my fancy button`
+          "text": `Thanks for the mention <@${message.user}>! Click my fancy button`
         },
         "accessory": {
           "type": "button",
@@ -132,41 +164,12 @@ app.event('app_mention', async ({ event, context, client, say }) => {
           "action_id": "first_button"
         }
       }
-    ]});
-  }
-  catch (error) {
-    console.error(error);
-  }
-});
-
-// subscribe to `message.channels` event in your App Config
-// need channels:read scope
-app.message('hello', async ({ message, say }) => {
-  // say() sends a message to the channel where the event was triggered
-  // no need to directly use 'chat.postMessage', no need to include token
-  await say({"blocks": [
-		{
-			"type": "section",
-			"text": {
-				"type": "mrkdwn",
-				"text": `Thanks for the mention <@${message.user}>! Click my fancy button`
-			},
-			"accessory": {
-				"type": "button",
-				"text": {
-					"type": "plain_text",
-					"text": "Button",
-					"emoji": true
-				},
-        "value": "click_me_123",
-        "action_id": "first_button"
-			}
-		}
-	]});
+    ]
+  });
 });
 
 // Listen and respond to button click
-app.action('first_button', async({action, ack, say, context}) => {
+app.action('first_button', async ({ action, ack, say, context }) => {
   console.log('button clicked');
   console.log(action);
   // acknowledge the request right away

--- a/examples/socket-mode/link.sh
+++ b/examples/socket-mode/link.sh
@@ -1,0 +1,11 @@
+#!/bin/bash
+
+current_dir=`dirname $0`
+cd ${current_dir}
+npm unlink @slack/bolt \
+  && npm i \
+  && cd ../.. \
+  && npm link \
+  && cd - \
+  && npm i \
+  && npm link @slack/bolt

--- a/src/index.ts
+++ b/src/index.ts
@@ -35,6 +35,11 @@ export {
 } from './receivers/HTTPModuleFunctions';
 export { HTTPResponseAck } from './receivers/HTTPResponseAck';
 
+export {
+  SocketModeFunctions,
+  SocketModeReceiverProcessEventErrorHandlerArgs,
+} from './receivers/SocketModeFunctions';
+
 export * from './errors';
 export * from './middleware/builtin';
 export * from './types';

--- a/src/receivers/HTTPModuleFunctions.ts
+++ b/src/receivers/HTTPModuleFunctions.ts
@@ -225,7 +225,7 @@ export interface RequestVerificationOptions {
   logger?: Logger;
 }
 
-// which handles errors occurred while dispatching a rqeuest
+// which handles errors occurred while dispatching a request
 export interface ReceiverDispatchErrorHandlerArgs {
   error: Error | CodedError;
   logger: Logger;

--- a/src/receivers/SocketModeFunctions.spec.ts
+++ b/src/receivers/SocketModeFunctions.spec.ts
@@ -1,0 +1,42 @@
+import 'mocha';
+import { assert } from 'chai';
+import { SocketModeFunctions as func } from './SocketModeFunctions';
+import {
+  ReceiverMultipleAckError,
+  AuthorizationError,
+} from '../errors';
+import { createFakeLogger } from '../test-helpers';
+import { ReceiverEvent } from '../types';
+
+describe('SocketModeFunctions', async () => {
+  describe('Error handlers for event processing', async () => {
+    const logger = createFakeLogger();
+
+    describe('defaultProcessEventErrorHandler', async () => {
+      it('should properly handle ReceiverMultipleAckError', async () => {
+        const event: ReceiverEvent = {
+          ack: async () => {},
+          body: {},
+        };
+        const shouldBeAcked = await func.defaultProcessEventErrorHandler({
+          error: new ReceiverMultipleAckError(),
+          logger,
+          event,
+        });
+        assert.isFalse(shouldBeAcked);
+      });
+      it('should properly handle AuthorizationError', async () => {
+        const event: ReceiverEvent = {
+          ack: async () => {},
+          body: {},
+        };
+        const shouldBeAcked = await func.defaultProcessEventErrorHandler({
+          error: new AuthorizationError('msg', new Error()),
+          logger,
+          event,
+        });
+        assert.isTrue(shouldBeAcked);
+      });
+    });
+  });
+});

--- a/src/receivers/SocketModeFunctions.spec.ts
+++ b/src/receivers/SocketModeFunctions.spec.ts
@@ -13,7 +13,7 @@ describe('SocketModeFunctions', async () => {
     const logger = createFakeLogger();
 
     describe('defaultProcessEventErrorHandler', async () => {
-      it('should properly handle ReceiverMultipleAckError', async () => {
+      it('should return false if passed any Error other than AuthorizationError', async () => {
         const event: ReceiverEvent = {
           ack: async () => {},
           body: {},
@@ -25,7 +25,7 @@ describe('SocketModeFunctions', async () => {
         });
         assert.isFalse(shouldBeAcked);
       });
-      it('should properly handle AuthorizationError', async () => {
+      it('should return true if passed an AuthorizationError', async () => {
         const event: ReceiverEvent = {
           ack: async () => {},
           body: {},

--- a/src/receivers/SocketModeFunctions.ts
+++ b/src/receivers/SocketModeFunctions.ts
@@ -17,7 +17,7 @@ export class SocketModeFunctions {
     // TODO: more details like envelop_id, payload type etc. here
     // To make them available, we need to enhance underlying SocketModeClient
     // to return more properties to 'slack_event' listeners
-    logger.error('An unhandled error occurred while Bolt processed an event');
+    logger.error(`An unhandled error occurred while Bolt processed a ${event.type} event: ${error}`);
     logger.debug(`Error details: ${error}, retry num: ${event.retryNum}, retry reason: ${event.retryReason}`);
     const errorCode = (error as CodedError).code;
     if (errorCode === ErrorCode.AuthorizationError) {

--- a/src/receivers/SocketModeFunctions.ts
+++ b/src/receivers/SocketModeFunctions.ts
@@ -17,7 +17,7 @@ export class SocketModeFunctions {
     // TODO: more details like envelop_id, payload type etc. here
     // To make them available, we need to enhance underlying SocketModeClient
     // to return more properties to 'slack_event' listeners
-    logger.error(`An unhandled error occurred while Bolt processed a ${event.type} event: ${error}`);
+    logger.error(`An unhandled error occurred while Bolt processed a ${event.body.type} event: ${error}`);
     logger.debug(`Error details: ${error}, retry num: ${event.retryNum}, retry reason: ${event.retryReason}`);
     const errorCode = (error as CodedError).code;
     if (errorCode === ErrorCode.AuthorizationError) {

--- a/src/receivers/SocketModeFunctions.ts
+++ b/src/receivers/SocketModeFunctions.ts
@@ -17,7 +17,7 @@ export class SocketModeFunctions {
     // TODO: more details like envelop_id, payload type etc. here
     // To make them available, we need to enhance underlying SocketModeClient
     // to return more properties to 'slack_event' listeners
-    logger.error(`An unhandled error occurred while Bolt processed a ${event.body.type} event: ${error}`);
+    logger.error(`An unhandled error occurred while Bolt processed (type: ${event.body.type}, error: ${error})`);
     logger.debug(`Error details: ${error}, retry num: ${event.retryNum}, retry reason: ${event.retryReason}`);
     const errorCode = (error as CodedError).code;
     if (errorCode === ErrorCode.AuthorizationError) {

--- a/src/receivers/SocketModeFunctions.ts
+++ b/src/receivers/SocketModeFunctions.ts
@@ -1,0 +1,38 @@
+/* eslint-disable import/prefer-default-export */
+import type { Logger } from '@slack/logger';
+import { CodedError, ErrorCode } from '../errors';
+import { ReceiverEvent } from '../types';
+
+export class SocketModeFunctions {
+  // ------------------------------------------
+  // Error handlers for event processing
+  // ------------------------------------------
+
+  // The default processEventErrorHandler implementation:
+  // Developers can customize this behavior by passing processEventErrorHandler to the constructor
+  public static async defaultProcessEventErrorHandler(
+    args: SocketModeReceiverProcessEventErrorHandlerArgs,
+  ): Promise<boolean> {
+    const { error, logger, event } = args;
+    // TODO: more details like envelop_id, payload type etc. here
+    // To make them available, we need to enhance underlying SocketModeClient
+    // to return more properties to 'slack_event' listeners
+    logger.error('An unhandled error occurred while Bolt processed an event');
+    logger.debug(`Error details: ${error}, retry num: ${event.retryNum}, retry reason: ${event.retryReason}`);
+    const errorCode = (error as CodedError).code;
+    if (errorCode === ErrorCode.AuthorizationError) {
+      // The `authorize` function threw an exception, which means there is no valid installation data.
+      // In this case, we can tell the Slack server-side to stop retries.
+      return true;
+    }
+    return false;
+  }
+}
+
+// The arguments for the processEventErrorHandler,
+// which handles errors `await app.processEvent(even)` method throws
+export interface SocketModeReceiverProcessEventErrorHandlerArgs {
+  error: Error | CodedError;
+  logger: Logger;
+  event: ReceiverEvent,
+}

--- a/src/receivers/SocketModeReceiver.ts
+++ b/src/receivers/SocketModeReceiver.ts
@@ -6,10 +6,15 @@ import { Logger, ConsoleLogger, LogLevel } from '@slack/logger';
 import { InstallProvider, CallbackOptions, InstallProviderOptions, InstallURLOptions, InstallPathOptions } from '@slack/oauth';
 import { AppsConnectionsOpenResponse } from '@slack/web-api';
 import App from '../App';
+import { CodedError } from '../errors';
 import { Receiver, ReceiverEvent } from '../types';
 import { StringIndexed } from '../types/helpers';
 import { buildReceiverRoutes, ReceiverRoutes } from './custom-routes';
 import { verifyRedirectOpts } from './verify-redirect-opts';
+import {
+  SocketModeFunctions as socketModeFunc,
+  SocketModeReceiverProcessEventErrorHandlerArgs,
+} from './SocketModeFunctions';
 
 // TODO: we throw away the key names for endpoints, so maybe we should use this interface. is it better for migrations?
 // if that's the reason, let's document that with a comment.
@@ -26,6 +31,7 @@ export interface SocketModeReceiverOptions {
   appToken: string; // App Level Token
   customRoutes?: CustomRoute[];
   customPropertiesExtractor?: (args: any) => StringIndexed;
+  processEventErrorHandler?: (args: SocketModeReceiverProcessEventErrorHandlerArgs) => Promise<boolean>;
 }
 
 export interface CustomRoute {
@@ -74,6 +80,8 @@ export default class SocketModeReceiver implements Receiver {
 
   private routes: ReceiverRoutes;
 
+  private processEventErrorHandler: (args: SocketModeReceiverProcessEventErrorHandlerArgs) => Promise<boolean>;
+
   public constructor({
     appToken,
     logger = undefined,
@@ -87,6 +95,7 @@ export default class SocketModeReceiver implements Receiver {
     installerOptions = {},
     customRoutes = [],
     customPropertiesExtractor = (_args) => ({}),
+    processEventErrorHandler = socketModeFunc.defaultProcessEventErrorHandler,
   }: SocketModeReceiverOptions) {
     this.client = new SocketModeClient({
       appToken,
@@ -101,6 +110,7 @@ export default class SocketModeReceiver implements Receiver {
       return defaultLogger;
     })();
     this.routes = buildReceiverRoutes(customRoutes);
+    this.processEventErrorHandler = processEventErrorHandler;
 
     // Verify redirect options if supplied, throws coded error if invalid
     verifyRedirectOpts({ redirectUri, redirectUriPath: installerOptions.redirectUriPath });
@@ -204,7 +214,18 @@ export default class SocketModeReceiver implements Receiver {
         retryReason: retry_reason,
         customProperties: customPropertiesExtractor(args),
       };
-      await this.app?.processEvent(event);
+      try {
+        await this.app?.processEvent(event);
+      } catch (error) {
+        const shouldBeAcked = await this.processEventErrorHandler({
+          error: error as Error | CodedError,
+          logger: this.logger,
+          event,
+        });
+        if (shouldBeAcked) {
+          await ack();
+        }
+      }
     });
   }
 

